google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITBlobReadChannelTest.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage.it;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.cloud.NoCredentials;
+import com.google.cloud.ReadChannel;
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.DataGeneration;
+import com.google.cloud.storage.LimitableReadChannel;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import com.google.cloud.storage.conformance.retry.TestBench;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Random;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestName;
+
+public final class ITBlobReadChannelTest {
+
+  private static final int _16MiB = 16 * 1024 * 1024;
+  private static final int _256KiB = 256 * 1024;
+
+  @ClassRule
+  public static final TestBench testBench =
+      TestBench.newBuilder().setContainerName("blob-read-channel-test").build();
+
+  @Rule public final TestName testName = new TestName();
+
+  @Rule public final DataGeneration dataGeneration = new DataGeneration(new Random(872364872));
+
+  @Test
+  public void testLimit_smallerThanOneChunk() throws IOException {
+    int srcContentSize = _256KiB;
+    int rangeBegin = 57;
+    int rangeEnd = 2384;
+    int chunkSize = _16MiB;
+    doLimitTest(srcContentSize, rangeBegin, rangeEnd, chunkSize);
+  }
+
+  @Test
+  public void testLimit_largerThanOneChunk() throws IOException {
+    int srcContentSize = _16MiB + (_256KiB * 3);
+    int rangeBegin = 384;
+    int rangeEnd = rangeBegin + _16MiB;
+    int chunkSize = _16MiB;
+
+    doLimitTest(srcContentSize, rangeBegin, rangeEnd, chunkSize);
+  }
+
+  private void doLimitTest(int srcContentSize, int rangeBegin, int rangeEnd, int chunkSize)
+      throws IOException {
+    Storage s =
+        StorageOptions.newBuilder()
+            .setProjectId("blob-read-channel-test")
+            .setHost(testBench.getBaseUri())
+            .setCredentials(NoCredentials.getInstance())
+            .build()
+            .getService();
+
+    Bucket bucket = s.create(BucketInfo.of("bucket"));
+    BlobInfo src =
+        BlobInfo.newBuilder(bucket, String.format("%s/src", testName.getMethodName())).build();
+    ByteBuffer content = dataGeneration.randByteBuffer(srcContentSize);
+    ByteBuffer expectedSubContent = content.duplicate();
+    expectedSubContent.position(rangeBegin);
+    expectedSubContent.limit(rangeEnd);
+    try (WriteChannel writer = s.writer(src)) {
+      writer.write(content);
+    }
+
+    ByteBuffer actual = ByteBuffer.allocate(rangeEnd - rangeBegin);
+
+    try (ReadChannel reader = s.reader(src.getBlobId())) {
```

Comment:
Investigate if it's possible (binary & source compatibility) to directly return, instead of down-casting.

Reply:
The alternative is to expose an alias that returns the new LimitableReadChannel class type, but based on discussion we didn't want to increase the surface of the API in this library so easily; which is fair.

Reply:
limit feature added to ReadChannel, no need for changes here.